apiserver/facades/agent/uniter/networkinfocaas.go

```diff
@@ -0,0 +1,208 @@
+		}
+		endpointIngressAddresses[endpoint] = ingress
+	}
+
+	// For CAAS units, we build up a minimal result struct
+	// based on the default space and unit public/private addresses,
+	// ie the addresses of the CAAS service.
+	addrs, err := n.unit.AllAddresses()
+	if err != nil {
+		return params.NetworkInfoResults{}, err
+	}
+	corenetwork.SortAddresses(addrs)
+
+	// We record the interface addresses as the machine local ones - these
+	// are used later as the binding addresses.
+	// For CAAS models, we need to default ingress addresses to all available
+	// addresses so record those in the default ingress address slice.
+	var interfaceAddr []network.InterfaceAddress
+	var defaultIngressAddresses []string
+	for _, a := range addrs {
+		if a.Scope == corenetwork.ScopeMachineLocal {
+			interfaceAddr = append(interfaceAddr, network.InterfaceAddress{Address: a.Value})
+		} else {
+			defaultIngressAddresses = append(defaultIngressAddresses, a.Value)
+		}
+	}
+
+	networkInfos := make(map[string]machineNetworkInfoResult)
+	networkInfos[corenetwork.AlphaSpaceId] = machineNetworkInfoResult{
+		NetworkInfos: []network.NetworkInfo{{Addresses: interfaceAddr}},
+	}
+
+	for endpoint, space := range bindings {
+		// The binding address information based on link layer devices.
+		info := machineNetworkInfoResultToNetworkInfoResult(networkInfos[space])
+
+		// Set egress and ingress address information.
+		info.EgressSubnets = endpointEgressSubnets[endpoint]
+
+		ingressAddrs := make([]string, len(endpointIngressAddresses[endpoint]))
+		for i, addr := range endpointIngressAddresses[endpoint] {
+			ingressAddrs[i] = addr.Value
+		}
+		info.IngressAddresses = ingressAddrs
+
+		// If there is no ingress address explicitly defined for a given
+		// binding, set the ingress addresses to either any defaults set above,
+		// or the binding addresses.
+		if len(info.IngressAddresses) == 0 {
+			info.IngressAddresses = defaultIngressAddresses
+		}
+
+		if len(info.IngressAddresses) == 0 {
+			ingress := spaceAddressesFromNetworkInfo(networkInfos[space].NetworkInfos)
+			corenetwork.SortAddresses(ingress)
+			info.IngressAddresses = make([]string, len(ingress))
+			for i, addr := range ingress {
+				info.IngressAddresses[i] = addr.Value
+			}
+		}
+
+		// If there is no egress subnet explicitly defined for a given binding,
+		// default to the first ingress address. This matches the behaviour when
+		// there's a relation in place.
+		if len(info.EgressSubnets) == 0 && len(info.IngressAddresses) > 0 {
+			var err error
+			info.EgressSubnets, err = network.FormatAsCIDR([]string{info.IngressAddresses[0]})
+			if err != nil {
+				return result, errors.Trace(err)
+			}
+		}
+
+		result.Results[endpoint] = info
+	}
+
+	return dedupNetworkInfoResults(result), nil
+}
+
+// getRelationNetworkInfo returns the endpoint name, network space
+// and ingress/egress addresses for the input relation ID.
+func (n *NetworkInfoCAAS) getRelationNetworkInfo(
```

Comment:
On the next pass, can we reduce the number of return arguments, it becomes meaningless after 3.

Reply:
Yes.